tariff/octopus: Properly support Export tariffs

duckfullstop opened the pull request:
This PR allows for setting `export: true` on a tariff template to acquire only the first Export tariff.

Works best in conjunction with the `feedin:` tariff type (obviously). Defaults to classic behaviour (which is to get the _import_ tariff).

duckfullstop committed:
tariff/octopus: Support setting whether to pull Export tariffs instead of just Import

duckfullstop commented:
As discussed in #23739 with @andig

duckfullstop committed:
tariff/octopus: Switch to using a tariffType choice template option

duckfullstop committed:
tariff/octopus: Don't unnecessarily throw warnings for tariffCode using Product Code mode

duckfullstop commented:
(apologies if this is total garbage, suffering some bad brain fog...)

duckfullstop committed:
tariff/octopus: Lint

duckfullstop committed:
tariff/octopus: Report type of tariff direction when unable to find appropriate tariffCode

duckfullstop committed:
tariff/octopus: Unify usage of tariffDirection

andig committed:
Update tariff/octopus.go

andig commented on tariff/octopus/graphql/types.go:
```diff
@@ -38,9 +38,22 @@ func (d *tariffData) TariffCode() string {
-func (d *tariffData) IsExport() bool {
-	return d.standardTariff.IsExport
+// TariffDirection defines which direction of energy flow is being denoted by the given tariff.
+type TariffDirection string
```
that's a little inconsistent. make it either tarifftype or direction, but don't use both

andig committed:
Update tariff/octopus/graphql/api.go

duckfullstop commented:
Apologies, been a bit under the weather the last week - I'll do my best to take a look later today 🙇

duckfullstop committed:
tariff/octopus: Stronger checking of tariffType

duckfullstop committed:
tariff/octopus: Standardise on tariffDirection nomenclature for energy flow direction

duckfullstop committed:
tariff/octopus: Permit "test" apiKey as allowed purely for testing purposes

Using a key that starts `sk_live` (even if it's obviously a test key) results in GitHub throwing secret protection faults.

duckfullstop committed:
tariff/octopus: Use new tariffDirection nomenclature in tests

duckfullstop committed:
tariff/octopus: Make tariff provider generation testable without API call

duckfullstop committed:
tariff/octopus: Handle product code configuration not providing TariffDirection

(sets a default on the TariffDirection to Import, which makes sense)

andig committed:
wip
